file/join.rs

```diff
@@ -0,0 +1,224 @@
+    #[arg(short = '2', long, default_value_t = 1)]
+    field2: usize,
+
+    /// File 1
+    file1: PathBuf,
+
+    /// File 2
+    file2: PathBuf,
+}
+
+fn parse_fields(line: &str, sep: char) -> Vec<String> {
+    line.split(sep)
+        .map(|s| s.to_string())
+        .collect()
+}
+
+fn read_file_lines(file_path: &PathBuf, sep: char) -> Result<Vec<Vec<String>>, Box<dyn std::error::Error>> {
+    let file = File::open(file_path).map_err(|_| format!("Unable to open file"))?;
+    let reader = BufReader::new(file);
+    let mut lines = Vec::new();
+
+    for line in reader.lines() {
+        let line = line.map_err(|_| format!("Unable to read line"))?;
+        lines.push(parse_fields(&line, sep));
+    }
+
+    Ok(lines)
+}
+
+fn perform_join(
+    file1: Vec<Vec<String>>,
+    file2: Vec<Vec<String>>,
+    field1: usize,
+    field2: usize,
+    a: Option<u8>,
+    e: Option<String>,
+    o: Option<Vec<String>>,
+    v: Option<u8>,
+) -> Result<(), Box<dyn std::error::Error>>{
+    if field1 == 0 || field2 == 0 {
+        return Err(format!("field 1 and field 2 must be greater than one").into());
+    }
+
+    if o.is_some() {
+        let mut i = 0;
+        for s in &file1 {
+            let mut res = Vec::new();
+            for num in o.clone().unwrap() {
+                let f_num: Vec<&str> = num.split('.').collect();
+                if f_num[0] == "1" {
+                    res.push(s[f_num[1].parse::<usize>().unwrap() - 1].clone());
+                }
+                if f_num[0] == "2" {
+                    if i >= file2.len() && e.is_some() {
+                        res.push(e.clone().unwrap());
+                    }
+                    else {
+                        res.push(file2[i][f_num[1].parse::<usize>().unwrap() - 1].clone());
+                    }
+                }
+            }
+            i += 1;
+            
+            println!("{}", res.join(" "));
+            if i >= file2.len() && e.is_none() {
+                break;
+            }
+        }
+        return Ok(());
+    }
+
+    let mut map: HashMap<String, Vec<Vec<String>>> = HashMap::new();
+
+    for line in &file1 {
+        let key = line[field1 - 1].clone();
+        map.entry(key).or_insert_with(Vec::new).push(line.clone());
+    }
+
+    let mut matched: HashMap<String, bool> = HashMap::new();
+
+    for line in &file2 {
+        let key = line[field2 - 1].clone();
+        if let Some(matches) = map.get_mut(&key) {
+            matched.insert(key.clone(), true);
+            for l in matches.iter_mut() {
+                let mut output = vec![];
+
+                output.extend_from_slice(&l);
+                output.extend_from_slice(&line[1..]);
+
+                if v.unwrap_or(0) != 1 && v.unwrap_or(0) != 2 {
+                    println!("{}", output.join(" "));
+                }
+            }
+        } else if a.unwrap_or(0) == 2 {
+            let mut output = vec![String::from("(unknown)")];
```

Comment:
Empty fields by default stay empty. `-e` is used to replace them with a custom filed.